alternator/ttl.cc

```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+// with this range), but when this node is down, other nodes may take over
+// (FIXME: this is not implemented yet).
+// An expiration thread is reponsible for all tables which need expiration
+// scans. FIXME: explain how this is done with multiple tables - parallel,
+// staggered, or what?
+// The expiration thread scans item using CL=QUORUM to ensures that it reads
+// a consistent expiration-time attribute. This means that the items are read
+// locally and in addition QUORUM-1 additional nodes (one additional node
+// when RF=3) need to read the data and send digests.
+// FIXME: explain if we can read the exact attribute or the entire map.
+// When the expiration thread decides that an item has expired and wants
+// to delete it, it does it using a CL=QUORUM write. This allows this
+// deletion to be visible for consistent (quorum) reads. The deletion,
+// like user deletions, will also appear on the CDC log and therefore
+// Alternator Streams if enabled (FIXME: explain how we mark the
+// deletion different from user deletes. We don't do it yet.).
+expiration_service::expiration_service(database& db, service::storage_proxy& proxy)
+        : _db(db)
+        , _proxy(proxy)
+{
+    //FIXME: add metrics for the service
+    //setup_metrics();
+}
+
+// The following is_expired() functions all check if an item with the given
+// expiration time has expired, according to the DynamoDB API rules.
+// The rules are:
+// 1. If the expiration time attribute's value is not a number type,
+//    the item is not expired.
+// 2. The expiration time is measured in seconds since the UNIX epoch.
+// 3. If the expiration time is more than 5 years in the past, it is assumed
+//    to be malformed and ignored - and the item does not expire.
+static bool is_expired(gc_clock::time_point expiration_time, gc_clock::time_point now) {
+    return expiration_time <= now &&
+           expiration_time > now - std::chrono::years(5);
+}
+static bool is_expired(const big_decimal& expiration_time, gc_clock::time_point now) {
+    // Convert the big_decimal to an integer. If it doesn't fit, the
+    // conversion will result in the highest number that does fit.
+    // FIXME: what is expiration_time.scale()? Can it ever be not 1?
+    // Add a test that tries 12345e1 as a timestamp!
+    uint32_t t = static_cast<uint32_t>(expiration_time.unscaled_value());
+    // We assume - and the assumption turns out to be correct - that the
+    // epoch of gc_clock::time_point and the one used by the DynamoDB protocol
+    // are the same (the UNIX epoch in UTC). The resolution (seconds) is also
+    // the same.
+    return is_expired(gc_clock::time_point(gc_clock::duration(std::chrono::seconds(t))), now);
+}
+static bool is_expired(const rjson::value& expiration_time, gc_clock::time_point now) {
+    std::optional<big_decimal> n = try_unwrap_number(expiration_time);
+    if (n) {
+        return is_expired(*n, now);
+    }
+    return false;
```

Comment:
nitpick: sounds like it can be folded to `return n && is_expired(*n, now)`

Reply:
Right :-) Will do it.